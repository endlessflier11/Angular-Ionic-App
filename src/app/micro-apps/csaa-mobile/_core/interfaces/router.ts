import { Route, UrlTree } from '@angular/router';
import { NavigationOptions } from '@ionic/angular/providers/nav-controller';
import { Observable } from 'rxjs';

interface AppRoute extends Route {
  name?: string;
  isGroupIndex?: boolean; // If true, it means route is the index of parent. Then it inherits all parent properties
}

export type AppRoutes = AppRoute[];

export interface RouteParam {
  [key: string]: string;
}

export interface NamedRouteConfig {
  name: string;
  path: string;
  children?: NamedRouteConfig[];
}

export interface NamedRoute {
  name: string;
  path: string;
  relativePath: string;
}

export interface CanComponentDeactivate {
  canDeactivate: () =>
    | Observable<boolean | UrlTree>
    | Promise<boolean | UrlTree>
    | boolean
    | UrlTree;
}

export interface CsaaRouteStackItem {
  name: string;
  params?: RouteParam;
  options?: NavigationOptions;
}
