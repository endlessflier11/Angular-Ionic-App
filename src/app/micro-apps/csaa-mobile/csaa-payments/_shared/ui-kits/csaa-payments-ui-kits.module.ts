import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { CommonModule } from '@angular/common';
import { CsaaPaymentHistoryCardComponent } from './csaa-payment-history-card/csaa-payment-history-card.component';
import { UiKitsModule } from '../../../_core/ui-kits/ui-kits.module';
import { CsaaPaymentHistorySelectCardComponent } from './csaa-payment-history-select-card/csaa-payment-history-select-card.component';
import { CsaaPaymentSucceededCardComponent } from './csaa-payment-succeeded-card/csaa-payment-succeeded-card.component';
import { CsaaMakePaymentCardComponent } from './csaa-make-payment-card/csaa-make-payment-card.component';
import { CsaaPayAllPoliciesCardComponent } from './csaa-pay-all-policies-card/csaa-pay-all-policies-card.component';
import { PaymentTermsConditionsComponent } from './payment-terms-conditions/payment-terms-conditions.component';
import { AutopayTermsConditionsComponent } from './autopay-terms-conditions/autopay-terms-conditions.component';
import { CsaaBillsCardComponent } from './csaa-bills-card/csaa-bills-card.component';
import { AutopayDiscountCardComponent } from './autopay-discount-card/autopay-discount-card.component';

const SHARED_COMPONENTS = [
  CsaaPaymentHistoryCardComponent,
  CsaaPaymentHistorySelectCardComponent,
  CsaaMakePaymentCardComponent,
  CsaaPaymentSucceededCardComponent,
  CsaaPayAllPoliciesCardComponent,
  PaymentTermsConditionsComponent,
  AutopayTermsConditionsComponent,
  CsaaBillsCardComponent,
  AutopayDiscountCardComponent,
];

@NgModule({
  declarations: [...SHARED_COMPONENTS],
  imports: [CommonModule, IonicModule, UiKitsModule, IonicModule],
  exports: [...SHARED_COMPONENTS],
})
export class CsaaPaymentsUiKitsModule {}
