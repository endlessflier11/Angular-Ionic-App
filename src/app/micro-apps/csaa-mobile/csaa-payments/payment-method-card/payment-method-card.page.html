<ion-header [ngClass]="['csaa', currentTheme]">
  <ion-toolbar>
    <ion-buttons slot="start">
      <csaa-back-button (back)="onClickBackBtn()" [text]=""></csaa-back-button>
    </ion-buttons>
    <ion-title>Pay with</ion-title>
    <ion-buttons slot="end">
      <ion-button (click)="done()" [disabled]="!canSubmitForm">Done</ion-button>
    </ion-buttons>
  </ion-toolbar>
</ion-header>
<ion-content [ngClass]="['csaa', currentTheme]" class="ion-text-center">
  <h1 class="ion-text-center">Credit or Debit Card</h1>
  <h2 class="ion-text-center">Enter Card Information</h2>

  <ion-button fill="clear" (click)="scanCard()" *ngIf="!isEditing" id="scan-card-button">
    <b>SCAN CARD</b>
  </ion-button>

  <form [formGroup]="cardDetailsForm">
    <div class="error-message" *ngIf="cardHolderName.invalid && cardHolderName.touched">
      <span *ngIf="cardHolderName.errors['required']">The name is required</span>
    </div>
    <ion-item lines="full">
      <ion-label position="fixed">Name</ion-label>
      <ion-input
        formControlName="cardHolderName"
        type="text"
        placeholder="Required"
        required
      ></ion-input>
    </ion-item>

    <div class="error-message" *ngIf="cardNumber.invalid && cardNumber.touched">
      <span *ngIf="cardNumber.errors['cardNumber'] && !cardNumber.errors['required']"
        >The card number is invalid</span
      >
      <span *ngIf="cardNumber.errors['required']">The card number is required</span>
    </div>
    <ion-item lines="full">
      <ion-label position="fixed">Card Number</ion-label>
      <ion-input
        formControlName="cardNumber"
        type="text"
        maxlength="23"
        placeholder="Required"
        [brmasker]="cardDetailsMaskConfig.cardNumber"
        (ionFocus)="setMaskCardNumber(false)"
        (ionBlur)="setMaskCardNumber(true)"
      >
      </ion-input>
    </ion-item>

    <div class="error-message" *ngIf="expiration.invalid && expiration.touched">
      <span *ngIf="expiration.errors['expiration'] && !expiration.errors['required']"
        >The expiration is invalid</span
      >
      <span *ngIf="expiration.errors['required']">The expiration is required</span>
    </div>
    <ion-item lines="full">
      <ion-label position="fixed">Expiration</ion-label>
      <ion-input
        formControlName="expiration"
        type="text"
        maxlength="5"
        placeholder="MM/YY"
        [brmasker]="cardDetailsMaskConfig.expiration"
        (ionFocus)="highlightCardExpired = false"
        [ngClass]="{'card-expired':highlightCardExpired}"
      >
      </ion-input>
    </ion-item>

    <div class="error-message" *ngIf="zipCode.invalid && zipCode.touched">
      <span *ngIf="zipCode.errors['zipCode'] && !zipCode.errors['required']"
        >The Zip Code has invalid format</span
      >
      <span *ngIf="zipCode.errors['required']">The Zip Code is required</span>
    </div>
    <ion-item lines="full">
      <ion-label position="fixed">Zip Code</ion-label>
      <ion-input
        formControlName="zipCode"
        type="text"
        maxlength="5"
        placeholder="Required"
        [brmasker]="cardDetailsMaskConfig.zip"
      >
      </ion-input>
    </ion-item>

    <ion-item lines="full" *ngIf="(hasRegId$|async) && !isEditing" id="save-card-toggle">
      <ion-label position="fixed">Save Card</ion-label>
      <ion-toggle formControlName="saveCard" [checked]="false" item-end></ion-toggle>
    </ion-item>

    <div class="error-message" *ngIf="showNickname && nickname.invalid && nickname.touched">
      <span *ngIf="nickname.errors['required']">The nickname is required</span>
    </div>
    <ion-item lines="full" *ngIf="showNickname">
      <ion-label position="fixed">Nickname</ion-label>
      <ion-input formControlName="nickname" type="text" placeholder="Name your card"></ion-input>
    </ion-item>
  </form>
  <p class="ion-padding">
    By selecting done you agree to the
    <span class="csaa-link" (click)="openPaymentTerms()">payment terms and conditions</span>.
  </p>
  <ion-button
    fill="clear"
    color="danger"
    *ngIf="isEditing"
    class="delete-card"
    [csaaDeleteConfirm]="onDeleteConfirmed"
    [deleteConfirmTarget]="editingPaymentMethod"
  >
    Delete Card
  </ion-button>
</ion-content>
