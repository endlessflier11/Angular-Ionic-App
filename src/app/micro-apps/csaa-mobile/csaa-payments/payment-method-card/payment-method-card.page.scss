@import '../../_scss/variables';

ion-content {
  --background: var(--ion-color-primary-contrast);
}

h2 {
  margin-bottom: 18px;
  font-weight: normal;
}

ion-item {
  --highlight-color-valid: var(--ion-color-primary);

  ion-label.label-fixed {
    width: 130px;
    min-width: 130px;
    color: $csaa-lighter-color;
  }
}

.error-message {
  text-align: left;
  background-color: white;
  color: $csaa-emergency-color;
  padding: 16px 16px 0;
}

ion-button {
  &:not(.in-toolbar) {
    &.delete-card {
      font-size: 17px;
      font-weight: normal;
    }
  }
}
