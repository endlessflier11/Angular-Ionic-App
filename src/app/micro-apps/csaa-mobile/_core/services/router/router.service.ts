import { Injectable } from '@angular/core';
import { NavigationCancel, NavigationEnd, Router } from '@angular/router';
import { NavController } from '@ionic/angular';
import { AnimationOptions, NavigationOptions } from '@ionic/angular/providers/nav-controller';
import { filter, map, pairwise, startWith, take } from 'rxjs/operators';

import { CsaaCommonModule } from '../../../csaa-core/csaa-common.module';
import { RouterHelpers } from '../../helpers';
import { Category, CsaaRouteStackItem, EventName, EventType, RouteParam } from '../../interfaces';
import { AnalyticsService } from '../analytics.service';
import { ROOT_NAME__INDEX } from '../../../constants';
import { Store } from '@ngxs/store';
import { ConfigState } from '../../store/states/config.state';

@Injectable({
  providedIn: CsaaCommonModule,
})
export class RouterService {
  private readonly helpers = RouterHelpers;
  private stack: CsaaRouteStackItem[] = [];

  constructor(
    private readonly router: Router,
    private readonly navController: NavController,
    private readonly analyticsService: AnalyticsService,
    private readonly store: Store
  ) {
    // - When we first load the app, the index page is loaded
    // and our RouterService is not aware of this navigation at this time.
    // So this is a hack to push the index route to the stack from the beginning.
    // Of course, this is under the assumption that our only entry point is the index root.
    // During deep linking or page reload (browser), it's likely that this will fail.
    // - To push the actual route (no assumption), we will need to refactor this stack
    // so that item.name is replaced by item.url since we can't correctly deduce the route name from the url.
    // However, we can check a known route-name (see RouterService.currentRouteIs()) against a url to see if it's a match.
    this.stack.push({ name: ROOT_NAME__INDEX });

    this.store
      .select(ConfigState.configLoaded)
      .pipe(
        filter((loaded) => !!loaded),
        take(1)
      )
      .subscribe(() => {
        // Track regular flow for authenticated users
        this.router.events
          .pipe(
            filter((event) => event instanceof NavigationEnd),
            startWith({ url: '' }),
            pairwise()
          )
          .subscribe(this.lurkIgTabTappedEvents.bind(this));
        // Track the flow for guest users blocked by CsaaAuthGuard.canActivate
        this.router.events
          .pipe(
            filter((event) => event instanceof NavigationCancel),
            map((event) => [{ url: router.url }, event])
          )
          .subscribe(this.lurkIgTabTappedEvents.bind(this));
        // Track page changes
        this.router.events
          .pipe(
            filter((event) => event instanceof NavigationEnd),
            filter((event: NavigationEnd) => this.isCsaaModulesRoute(event.url))
          )
          .subscribe(() => this.analyticsService.page());
      });
  }

  /**
   * Track Insurance Tab Tapped event deriving the info from the route state
   * Will track everytime previous is from outside our module and actual going inside our module
   *
   * @param previous the rote user is coming from
   * @param actual the route user is going to
   */
  private lurkIgTabTappedEvents([previous, actual]: [
    NavigationEnd,
    NavigationEnd | NavigationCancel
  ]) {
    if (
      actual.url.includes(this.fullPath(ROOT_NAME__INDEX)) &&
      !previous.url.includes(this.getModuleRootPath())
    ) {
      this.analyticsService.trackEvent(EventName.INSURANCE_TAB_TAPPED, Category.application, {
        event_type: EventType.LINK_ACCESSED,
      });
    }
  }

  fullPath(name, params?: RouteParam) {
    return this.getModuleRootPath() + this.helpers.getRoutePath(name, params);
  }

  relativePath(name) {
    return this.helpers.getRouteRelativePath(name);
  }

  isCsaaModulesRoute(url: string = null) {
    const route = url || this.router.url;
    return !!route && route.startsWith(this.getModuleRootPath());
  }

  currentRouteIs(name, params?: RouteParam): boolean {
    const currentPath = this.router.url;
    const compiledPath = this.fullPath(name, params);
    return currentPath === compiledPath;
  }

  previousRouteIs(name, params?: RouteParam): boolean {
    if (this.stack.length < 2) {
      return false;
    }
    const lastItem = this.stack[this.stack.length - 2];
    return this.fullPath(lastItem.name, lastItem.params) === this.fullPath(name, params);
  }

  currentRouteContains(segment): boolean {
    return this.router.url.indexOf(segment) !== -1;
  }

  public getController(): NavController {
    return this.navController;
  }

  public navigateForward(
    name: string,
    params?: RouteParam,
    options?: NavigationOptions
  ): Promise<boolean> {
    if (name === ROOT_NAME__INDEX) {
      return this.navigateRoot(name, params, options);
    }
    this.stack.push({ name, params, options });
    return this.navController.navigateForward(this.fullPath(name, params), options);
  }

  public navigateBack(
    name: string,
    params?: RouteParam,
    options?: NavigationOptions
  ): Promise<boolean> {
    if (name === ROOT_NAME__INDEX) {
      return this.navigateRoot(name, params, options);
    }

    let to = this.stack.pop();
    while (to && to.name !== name) {
      to = this.stack.pop();
    }
    this.stack.pop();
    this.stack.push({ name, params, options });
    return this.navController.navigateBack(this.fullPath(name, params), options);
  }

  public navigateRoot(
    name: string = ROOT_NAME__INDEX,
    params?: RouteParam,
    options?: NavigationOptions
  ): Promise<boolean> {
    this.stack = [{ name, params, options }];
    return this.navController.navigateRoot(this.fullPath(name, params), options);
  }

  public back(options?: AnimationOptions): void {
    this.stack.pop();
    let to = this.stack.pop();
    if (!to) {
      to = { name: ROOT_NAME__INDEX };
    }
    to.options = options;
    this.stack.push(to);
    if (to.name === ROOT_NAME__INDEX) {
      this.navigateRoot(to.name, to.params, to.options);
    } else {
      this.navController.navigateBack(this.fullPath(to.name, to.params), to.options);
    }
  }

  public getTop(): CsaaRouteStackItem {
    if (this.stack.length > 0) {
      return { ...this.stack[this.stack.length - 1] };
    }
    return { name: ROOT_NAME__INDEX };
  }

  public navigateAway(path: string) {
    this.stack = [{ name: ROOT_NAME__INDEX }];
    this.navController.navigateBack(path);
  }

  private getModuleRootPath() {
    const { moduleRootPath } = this.store.selectSnapshot(ConfigState);
    return moduleRootPath;
  }

  backToClubHome() {
    const { homeBackButtonRedirectTo, moduleRootPath } = this.store.selectSnapshot(ConfigState);
    if (this.router.url !== moduleRootPath) {
      // If the user is not at home page, make sure to navigate back to HOME
      // due to MWG tabbed design and the way ionic tab works with router
      this.navigateRoot();
      setTimeout(() => this.navigateAway(homeBackButtonRedirectTo));
    } else {
      this.navigateAway(homeBackButtonRedirectTo);
    }
  }
}
