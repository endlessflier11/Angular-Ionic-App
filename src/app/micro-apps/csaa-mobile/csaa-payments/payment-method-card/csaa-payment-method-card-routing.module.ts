import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AppRoutes } from '../../_core/interfaces';
import { PaymentMethodCardPage } from './payment-method-card.page';

const routes: AppRoutes = [
  {
    path: '',
    isGroupIndex: true,
    component: PaymentMethodCardPage,
  },
];

@NgModule({
  declarations: [],
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class CsaaPaymentMethodCardRoutingModule {}
