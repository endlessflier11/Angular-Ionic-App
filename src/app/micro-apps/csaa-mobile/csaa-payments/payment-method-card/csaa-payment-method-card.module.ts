import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CsaaPaymentMethodCardRoutingModule } from './csaa-payment-method-card-routing.module';
import { PaymentMethodCardPage } from './payment-method-card.page';
import { IonicModule } from '@ionic/angular';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';
import { ReactiveFormsModule } from '@angular/forms';
import { BrMaskerModule } from 'br-mask';
import { CsaaPaymentDirectivesModule } from '../_shared/directives/csaa-payment-directives.module';

@NgModule({
  declarations: [PaymentMethodCardPage],
  imports: [
    CommonModule,
    CsaaPaymentMethodCardRoutingModule,
    ReactiveFormsModule,
    IonicModule,
    UiKitsModule,
    BrMaskerModule,
    CsaaPaymentDirectivesModule,
  ],
})
export class CsaaPaymentMethodCardModule {}
