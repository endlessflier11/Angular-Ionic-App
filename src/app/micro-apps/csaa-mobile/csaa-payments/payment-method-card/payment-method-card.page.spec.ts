import { ComponentFixture, TestBed } from '@angular/core/testing';
import { IonicModule, LoadingController, Platform } from '@ionic/angular';
import { PaymentMethodCardPage } from './payment-method-card.page';
import { PageTestingModule, By, StoreTestBuilder } from '@app/testing';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';
import { CardIO } from '@ionic-native/card-io/ngx';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { AnalyticsService } from '../../_core/services/analytics.service';
import { PaymentService } from '../../_core/services/payment.service';
import { of } from 'rxjs';
import { MakePaymentService } from '../_shared/services/make-payment.service';
import { ActivatedRoute } from '@angular/router';
import { CallService } from '../../_core/services/call.service';
import { GlobalStateService } from '../../_core/services/global-state.service';
import { CsaaAppInjector } from '../../csaa-app.injector';
import { Injector } from '@angular/core';

import { delay } from '../../_core/helpers/async.helper';
import { CsaaPaymentDirectivesModule } from '../_shared/directives/csaa-payment-directives.module';
import { Category, EventName } from '../../_core/interfaces/analytics.interface';
import { Store } from '@ngxs/store';
import { BrMaskDirective } from 'br-mask';

try {
  describe('PaymentMethodCardPage', () => {
    let analyticService;
    let component: PaymentMethodCardPage;
    let fixture: ComponentFixture<PaymentMethodCardPage>;
    const PARAM_MAP = new Map();
    PARAM_MAP.set('account', '1234');

    beforeEach(async () => {
      await TestBed.configureTestingModule({
        declarations: [PaymentMethodCardPage, BrMaskDirective],
        providers: [
          CardIO,

          {
            provide: PaymentService,
            useValue: {
              getPaymentHistory: jest.fn().mockReturnValue(of({})),

              getWallets: jest.fn().mockReturnValue(of({ paymentAccounts: [] })),
              refreshPayments: jest.fn(),
              refreshHistory: jest.fn(),
              getUpcomingPayments: jest.fn().mockReturnValue(of([])),
              refreshAll: jest.fn(),
            },
          },
          {
            provide: MakePaymentService,
            useValue: {
              setActivePayment: jest.fn(),
              getWalletData: jest.fn(),
            },
          },
          {
            provide: ActivatedRoute,
            useValue: {
              params: of({ autoPay: false, policyNumber: '12345' }),
              paramMap: of(PARAM_MAP),
              queryParams: of(PARAM_MAP),
            },
          },
          { provide: CallService, useValue: { call: jest.fn() } },
          {
            provide: GlobalStateService,
            useValue: {
              getIsStandalone: jest.fn(() => false),
              getRegistrationId: jest.fn().mockReturnValue('12345'),
            },
          },
          {
            provide: LoadingController,
            useValue: { create: jest.fn(() => ({ present: jest.fn() })) },
          },
        ],
        imports: [
          PageTestingModule.withConfig({
            providesStorage: true,
            providesConfig: true,
            providesAlert: true,
            providesAnalytics: true,
            providesRouter: true,
            providesModal: true,
          }),
          UiKitsModule,
          ReactiveFormsModule,
          IonicModule,
          RouterTestingModule,
          CsaaPaymentDirectivesModule,
        ],
      }).compileComponents();

      CsaaAppInjector.injector = TestBed.inject(Injector);

      const store = TestBed.inject(Store);
      StoreTestBuilder.withDefaultMocks().resetStateOf(store);
      await TestBed.inject(Platform).ready();

      analyticService = TestBed.inject(AnalyticsService);
      fixture = TestBed.createComponent(PaymentMethodCardPage);
      component = fixture.componentInstance;
      fixture.detectChanges();
    });

    it('should create', () => {
      expect(component).toBeTruthy();
    });

    it('should match snapshots when clean', async () => {
      await delay(50);
      expect(fixture).toMatchSnapshot();
    });

    it('should display page title for cards', () => {
      const h1 = fixture.debugElement.query(By.cssAndText('h1', 'Credit or Debit Card'));
      expect(h1).toBeTruthy();
    });

    it('should have Done button disabled when form is invalid', () => {
      const doneButton = fixture.nativeElement.querySelector(
        'ion-header ion-buttons[slot=end] ion-button'
      );
      expect(doneButton.disabled).toBe(true);
    });

    it('should have Done button enabled when form is valid', () => {
      const nameInput = fixture.nativeElement.querySelector('[formcontrolname=cardHolderName]');
      nameInput.value = 'John Wick';
      nameInput.dispatchEvent(new Event('ionChange'));
      const cardNumberInput = fixture.nativeElement.querySelector('[formcontrolname=cardNumber]');
      cardNumberInput.value = '5262669012548834';
      cardNumberInput.dispatchEvent(new Event('ionBlur'));
      cardNumberInput.dispatchEvent(new Event('ionChange'));
      const expirationInput = fixture.nativeElement.querySelector('[formcontrolname=expiration]');
      const validExpire = `12/${new Date().getFullYear() - 2000 + 1}`;
      expirationInput.value = validExpire;
      expirationInput.dispatchEvent(new Event('ionChange'));
      const zipCodeInput = fixture.nativeElement.querySelector('[formcontrolname=zipCode]');
      zipCodeInput.value = '12345';
      zipCodeInput.dispatchEvent(new Event('ionChange'));
      const nickName = fixture.nativeElement.querySelector('[formcontrolname=nickname]');
      nickName.value = 'nickname 123';
      nickName.dispatchEvent(new Event('ionChange'));
      fixture.detectChanges();
      const doneButton = fixture.nativeElement.querySelector(
        'ion-header ion-buttons[slot=end] ion-button'
      );
      expect(doneButton.disabled).toBe(false);
    });

    it('should show validation error when name is empty', () => {
      const nameInput = fixture.nativeElement.querySelector('[formcontrolname=cardHolderName]');
      nameInput.value = '';
      nameInput.dispatchEvent(new Event('ionFocus'));
      nameInput.dispatchEvent(new Event('ionBlur'));
      fixture.detectChanges();

      const validationMessage = fixture.debugElement.query(
        By.cssAndText('span', 'The name is required')
      );
      expect(validationMessage).toBeTruthy();
    });

    it('should show validation error when card number has incorrect format', () => {
      const cardNumberInput = fixture.nativeElement.querySelector('[formcontrolname=cardNumber]');
      cardNumberInput.dispatchEvent(new Event('ionFocus'));
      cardNumberInput.value = '123';
      cardNumberInput.dispatchEvent(new Event('ionChange'));
      cardNumberInput.dispatchEvent(new Event('ionBlur'));
      fixture.detectChanges();
      const validationMessage = fixture.debugElement.query(
        By.cssAndText('.error-message', 'The card number is invalid')
      );
      expect(validationMessage).toBeTruthy();
    });

    it('should show validation error when zip code has incorrect format', () => {
      const zipCodeInput = fixture.nativeElement.querySelector('[formcontrolname=zipCode]');
      zipCodeInput.dispatchEvent(new Event('ionFocus'));
      zipCodeInput.value = '1122';
      zipCodeInput.dispatchEvent(new Event('ionChange'));
      zipCodeInput.dispatchEvent(new Event('ionBlur'));

      fixture.detectChanges();

      const validationMessage = fixture.debugElement.query(
        By.cssAndText('.error-message', 'The Zip Code has invalid format')
      );
      expect(validationMessage).toBeTruthy();
    });

    it('should show delete button if received edit: true from navParams', () => {
      component.isEditing = true;
      fixture.detectChanges();

      const button = fixture.debugElement.query(By.css('.delete-card'));
      expect(button).toBeTruthy();
    });

    it('should hide scan card button if editing card', () => {
      component.isEditing = true;
      fixture.detectChanges();

      const button = fixture.debugElement.query(By.css('#scan-card-button'));
      expect(button).toBeFalsy();
    });

    it('should show saveCard toggle if not editing', () => {
      fixture.detectChanges();
      expect(fixture.debugElement.query(By.css('#save-card-toggle'))).toBeTruthy();
    });

    it('should track the open payment terms', async () => {
      await component.openPaymentTerms();
      expect(analyticService.trackEvent).toHaveBeenCalledWith(
        EventName.TERMS_AND_CONDITIONS_CLICKED,
        Category.payments,
        {
          event_type: 'File Downloaded',
        }
      );
    });

    it('should navigate to terms and conditions page when button is clicked', () => {
      const span = fixture.debugElement.query(
        By.cssAndText('span', 'payment terms and conditions')
      );
      expect(span).toBeTruthy();
    });
  });
} catch (error) {
  console.error(error);
  throw error;
}
