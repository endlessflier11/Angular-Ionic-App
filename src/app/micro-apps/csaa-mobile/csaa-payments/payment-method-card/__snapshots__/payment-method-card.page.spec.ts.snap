// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentMethodCardPage should match snapshots when clean 1`] = `
<app-csaa-payment-method-card
  __hasRegId$__selector={[Function Observable]}
  __ngContext__={[Function LRootView]}
  alertCtrl={[Function AlertControllerMock]}
  analyticsService={[Function AnalyticsMockService]}
  callService={[Function Object]}
  cardDetailsData={[Function Object]}
  cardDetailsForm={[Function FormGroup]}
  cardDetailsMaskConfig={[Function Object]}
  cardIO={[Function CardIO]}
  configService={[Function ConfigMockService]}
  currentTheme=""
  deleteConfirmDirective="undefined"
  editingPaymentMethod="undefined"
  formBuilder={[Function FormBuilder]}
  highlightCardExpired="false"
  isAutoPayFlow="false"
  isEditing="false"
  loading="false"
  loadingCtrl={[Function Object]}
  makePaymentService={[Function Object]}
  modalCtrl={[Function AlertControllerMock]}
  notInitialized={[Function Boolean]}
  onDeleteConfirmed={[Function Function]}
  paymentService={[Function Object]}
  route={[Function Object]}
  routerService={[Function RouterMockService]}
  store={[Function Store]}
  tearDown$={[Function Subject]}
  wallet={[Function Object]}
>
  <ion-header
    class="csaa"
  >
    <ion-toolbar>
      <ion-buttons
        slot="start"
      >
        <csaa-back-button>
          <ion-button>
            <ion-icon
              name="chevron-back-outline"
              slot="start"
            />
             Back
          </ion-button>
        </csaa-back-button>
      </ion-buttons>
      <ion-title>
        Pay with
      </ion-title>
      <ion-buttons
        slot="end"
      >
        <ion-button>
          Done
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header><ion-content
    class="ion-text-center csaa"
  >
    <h1
      class="ion-text-center"
    >
      Credit or Debit Card
    </h1>
    <h2
      class="ion-text-center"
    >
      Enter Card Information
    </h2>
    <ion-button
      fill="clear"
      id="scan-card-button"
    >
      <b>
        SCAN CARD
      </b>
    </ion-button>
    <form
      class="ng-untouched ng-pristine ng-invalid"
      novalidate=""
    >
      <ion-item
        class="ion-untouched ion-pristine ion-invalid"
        lines="full"
      >
        <ion-label
          position="fixed"
        >
          Name
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
          formcontrolname="cardHolderName"
          placeholder="Required"
          required=""
          type="text"
        />
      </ion-item>
      <ion-item
        class="ion-untouched ion-pristine ion-invalid"
        lines="full"
      >
        <ion-label
          position="fixed"
        >
          Card Number
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
          formcontrolname="cardNumber"
          maxlength="23"
          placeholder="Required"
          type="text"
        />
      </ion-item>
      <ion-item
        class="ion-untouched ion-pristine ion-invalid"
        lines="full"
      >
        <ion-label
          position="fixed"
        >
          Expiration
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
          formcontrolname="expiration"
          maxlength="5"
          placeholder="MM/YY"
          type="text"
        />
      </ion-item>
      <ion-item
        class="ion-untouched ion-pristine ion-invalid"
        lines="full"
      >
        <ion-label
          position="fixed"
        >
          Zip Code
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
          formcontrolname="zipCode"
          maxlength="5"
          placeholder="Required"
          type="text"
        />
      </ion-item>
      <ion-item
        class="ion-untouched ion-pristine ion-valid"
        id="save-card-toggle"
        lines="full"
      >
        <ion-label
          position="fixed"
        >
          Save Card
        </ion-label>
        <ion-toggle
          class="ng-untouched ng-pristine ng-valid ion-untouched ion-pristine ion-valid"
          formcontrolname="saveCard"
          item-end=""
        />
      </ion-item>
      <ion-item
        class="ion-untouched ion-pristine ion-valid"
        lines="full"
      >
        <ion-label
          position="fixed"
        >
          Nickname
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-valid ion-untouched ion-pristine ion-valid"
          formcontrolname="nickname"
          placeholder="Name your card"
          type="text"
        />
      </ion-item>
    </form>
    <p
      class="ion-padding"
    >
       By selecting done you agree to the 
      <span
        class="csaa-link"
      >
        payment terms and conditions
      </span>
      . 
    </p>
  </ion-content>
</app-csaa-payment-method-card>
`;
